src/kernel/include/thd.h

```diff
@@ -500,16 +497,106 @@ thd_invstk_push(struct thread *thd, struct comp_info *ci, unsigned long ip, unsi
 	prev->sp = sp;
 	memcpy(&top->comp_info, ci, sizeof(struct comp_info));
 	top->ip = top->sp = 0;
+	
+	if (likely(thd->ulk_invstk)) top->ulk_stkoff = thd->ulk_invstk->top;
 
 	return 0;
 }
 
+static inline int
+thd_invstk_push_ulk(struct thread *thd, struct comp_info *ci, struct cos_cpu_local_info *cos_info)
+{
+	struct invstk_entry *top;
+
+	if (unlikely(curr_invstk_top(cos_info) >= THD_INVSTK_MAXSZ)) return -1;
+
+	top  = &thd->invstk[curr_invstk_top(cos_info) + 1];
+	curr_invstk_inc(cos_info);
+	memcpy(&top->comp_info, ci, sizeof(struct comp_info));
+	
+	if (likely(thd->ulk_invstk)) top->ulk_stkoff = THD_UL_INV_ORIGIN;
+
+	return 0;
+}
+
+static inline void
+thd_invstk_pop_ulk(struct thread *thd, struct cos_cpu_local_info *cos_info)
+{
+	struct invstk_entry *curr;
+
+	if (unlikely(curr_invstk_top(cos_info) == 0)) return;
```

Comment:
If the top is already calculated in teh calling `syscall` function, then just pass it in. Redundant computations on the fastpath here aren't great, and it would be good if this function came down to just the conditional below. If not that, then maybe conditionally call this if there was actually a `push` (which we can track in a varaible in the calling function).